Assets/MixedRealityToolkit.Services/InputAnimation/InputAnimation.cs

```diff
@@ -0,0 +1,710 @@
+        {
+            public int Compare(InputAnimationMarker a, InputAnimationMarker b)
+            {
+                return a.time.CompareTo(b.time);
+            }
+        }
+
+        public InputAnimation()
+        {
+            handTrackedCurveLeft = new AnimationCurve();
+            handTrackedCurveRight = new AnimationCurve();
+            handPinchCurveLeft = new AnimationCurve();
+            handPinchCurveRight = new AnimationCurve();
+            handJointCurvesLeft = new Dictionary<TrackedHandJoint, PoseCurves>();
+            handJointCurvesRight = new Dictionary<TrackedHandJoint, PoseCurves>();
+            cameraCurves = new PoseCurves();
+
+            markers = new List<InputAnimationMarker>();
+        }
+
+        public bool TryGetHandJointCurves(Handedness handedness, TrackedHandJoint joint, out PoseCurves curves)
+        {
+            if (handedness == Handedness.Left)
+            {
+                return handJointCurvesLeft.TryGetValue(joint, out curves);
+            }
+            else if (handedness == Handedness.Right)
+            {
+                return handJointCurvesRight.TryGetValue(joint, out curves);
+            }
+            curves = new PoseCurves();
+            return false;
+        }
+
+        public PoseCurves CreateHandJointCurves(Handedness handedness, TrackedHandJoint joint)
+        {
+            if (handedness == Handedness.Left)
+            {
+                if (!handJointCurvesLeft.TryGetValue(joint, out PoseCurves curves))
+                {
+                    curves = new PoseCurves();
+                    handJointCurvesLeft.Add(joint, curves);
+                }
+                return curves;
+            }
+            else if (handedness == Handedness.Right)
+            {
+                if (!handJointCurvesRight.TryGetValue(joint, out PoseCurves curves))
+                {
+                    curves = new PoseCurves();
+                    handJointCurvesRight.Add(joint, curves);
+                }
+                return curves;
+            }
+            return null;
+        }
+
+        /// <summary>
+        /// Add a keyframe for the tracking state of a hand.
+        /// </summary>
+        public void AddHandStateKey(float time, Handedness handedness, bool isTracked, bool isPinching)
+        {
+            if (handedness == Handedness.Left)
+            {
+                AddHandStateKey(time, isTracked, isPinching, handTrackedCurveLeft, handPinchCurveLeft);
+            }
+            else if (handedness == Handedness.Right)
+            {
+                AddHandStateKey(time, isTracked, isPinching, handTrackedCurveRight, handPinchCurveRight);
+            }
+        }
+
+        /// <summary>
+        /// Add a keyframe for one hand joint.
+        /// </summary>
+        public void AddHandJointKey(float time, Handedness handedness, TrackedHandJoint joint, MixedRealityPose jointPose, float positionThreshold, float rotationThreshold)
```

Comment:
Nit: Add a comment about what the threshold is (it's a squared magnitude thing)

Reply:
Improved comments.

The squaring is just a minor internal performance thing to avoid square root in vector magnitude.